Cache

jvangestel opened the pull request:
This Pull request fixes Cache issues with Zend Framework 1 in 3 ways:


1. ( 7939aadd8c055cccb201ac81614b44b70a21292a ) Gems\Cache\Backend\APC and Gems\Cache\Backend\File   Fix the PHP 7.2 deprecation of the each function by replacing it with a foreach. As far as I could see this is the only use of each or other deprecated functions in use in GemsTracker and should enable us to run Gemstracker on PHP 7.2.

2. ( 9f3d802b31f300cad0eeb1d8fa07495e00b44935 ) Gems\Cache\Backend\Psr6Cache and Gems\Cache\Backend\Psr16Cache implement the PHP-FIG cache standards for [Cache](https://www.php-fig.org/psr/psr-6/) and [SimpleCache](https://www.php-fig.org/psr/psr-16/). SimpleCache is a simplified cache without some functions like tag support. These Zend Cache backends are basically translators enabling to add any complient caching library, e.g. Symfony Cache. The benifit of this is that the adapters written for these standards should be complient with any cache library, thus making it a lot easier to support more caching adapters.
In Zend Expressive we could create a psr-6 caching library, add that to the servicemanager, then put the same cache in this adapter and load it as legacyCache, thus enabling us to use the legacy cache, but also write for the new cache.
By the time we have updated all code that uses Zend_Cache (also used in zend_db and zend_translator!) we can remove this. 

3. ( 2e914f8f30a0b2067a781dea97662c6316fad7d9 ) Gems\Cache\Backend\ZendCache is a specific backend that makes it possible to add Zend Cache 2 to Zend Cache 1. If we decide to stay with Zend as cache library supplier, this would be the way to go. Zend Cache is not PSR-6 complient, though it has an adapter for it. This implementation is not finished, quite a few methods are still empty, but it could serve as a start if the need is there to directly use Zend Cache

4. ( 8950af37ba4538ca40d7a1f418b183193404f2e4 ) This is an example of how Symfony Cache should be implemented in GemsEscort. For now we will use Symfony Cache 3.4, as it has LTS support until nov 2020, and Symfony 4.x versions require 7.1 or higher. The File Cache seems to work, I have not been able to try the APC cache. Small note is that Psr-16 has been removed from symfony cache since version 4.2.

At the very least I'd like to add the first to 1.8.7. Though the second would be good to add as well, making it possible to implement in projects until a required 1.8.8?

Lets discuss this first before actually merging!

jvangestel commented:
I agree that further testing is needed for option (2+) 4 before deploying it in production environments.
I will later add 1 to the master branch including the GemsEscort changes to reference those classes. 
I'll also add the files in 2 for testing purposes. Do you have an idea how we can easily document the changes made in 4, so it'll be easy to switch it out?

mddejong commented:
As far as I am concerned this looks fine. I advise to switch immediately and change the code so that we work with PHP 7.2 compatible cache by default. Also in New project.

If a party wants to upgrade and keep using the old cache we'll see it then.

mddejong commented:
As far as I am concerned we should merge this pull request. 

I just had to implement this code for the AnimeCon Ticketbooth as someone installed PHP 7.3 there and it works fine.

jvangestel commented:
@mddejong you tested with the Symfony cache version, or the zend_cache fixes in Gems\Cache\Backend\File

I've fixed the Zend 2 adapter for a proper comparison. It now works as well.

I've decided to check performance. Spoiler: Symfony file cache is a bit slower, but probably not too bad.

```PHP
?php

class CacheTestController extends \Gems_Controller_Action
{

    /**
     * @var \Zend_Cache_Core
     */
    public $cache;


    public function indexAction()
    {
        $this->initRawOutput();

        $iterations = 1000;
        $i = 0;
        $data = [
            1,
            2,
            'hello',
            ['subarray!'],
            'test' => 0,
        ];

        $timeStart = microtime(true);

        while ($i<$iterations) {
            $i++;

            $data['test'] = $i;

            $this->cache->save($data, 'test');

            $loadedData = $this->cache->load('test');

            $this->cache->remove('test');

            //print_r($loadedData);

        }

        $timeEnd = microtime(true);

        echo 'Total execution time in seconds: ' . ($timeEnd - $timeStart) . "\n";
    }
}
```

I've run this with all 3 cache systems.
It's a VM running on a conventional HDD. So the actual times are not as representative as on actual servers.
I have tested multiple times, but the times mentioned are pretty representative.
1000 iterations gives around: 

Zend_Cache_Backend_File: Total execution time in seconds: 18.301223039627

Gems\Cache\Backend\Psr6Cache with \Symfony\Component\Cache\Adapter\FilesystemAdapter: 
Total execution time in seconds: 24.504446029663
\Symfony\Component\Cache\Adapter\FilesystemAdapter (without zend_cache wrapper): 
Total execution time in seconds: 21.871017932892

Gems\Cache\Backend\ZendCache with \Zend\Cache\Storage\Adapter\Filesystem: 
Total execution time in seconds: 13.449715137482
\Zend\Cache\Storage\Adapter\Filesystem (without zend_cache wrapper): 
Total execution time in seconds: 12.526371002197


So Symfony cache is 0.007s slower per Cache save/load than Zend_Cache
Zend2 is .005s faster per Cache save/load than Zend_Cache
Zend2 and Symfony differ 0.012s or close to 50%.

Percentual the results seem the same if cache results are only fetched, instead of written and deleted (around 10% of the above times)

Again, I have tested none of the caching with APC cache.

- By mere speed of the filesystem cache I'd say Going to Zend2 cache is the winner.
- On the other hand Symfony cache does seem nicer to work with. Easier and following standards.
- Both are actively supported
- Symfony cache has a tad higher code coverage.

In the end the difference isn't that big on actual application performance time (how many cache saves or loads does it actually do? initially the db tables, but after that?

My proposed solution for 1.8.7 would be to make \Zend_Cache solutions oldFile and oldApc, and symfony or Zend newFile and newApc. 
We default the current 'file' to 'oldFile'. 
We are going to set a couple of projects to 'newFile' (Pulse at least) to test it out. In 1.8.8 we default the current 'file' to 'newFile'.
